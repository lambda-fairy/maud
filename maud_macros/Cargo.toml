[package]
name = "maud_macros"
# When releasing a new version, please update html_root_url in src/lib.rs
version = "0.26.0"
documentation = "https://docs.rs/maud_macros/"
description = "Internal crate. Please use 'maud' instead."

authors.workspace = true
license.workspace = true
homepage.workspace = true
repository.workspace = true
edition.workspace = true

[dependencies]
syn = "2"
quote = "1.0.7"
proc-macro2 = "1.0.23"
proc-macro-error2 = { version = "2.0.1", default-features = false }

[lib]
name = "maud_macros"
proc-macro = true
