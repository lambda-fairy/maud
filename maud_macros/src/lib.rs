#![doc(html_root_url = "https://docs.rs/maud_macros/0.22.1")]
// TokenStream values are reference counted, and the mental overhead of tracking
// lifetimes outweighs the marginal gains from explicit borrowing
#![allow(clippy::needless_pass_by_value)]

extern crate proc_macro;

mod ast;
mod generate;
mod parse;

use proc_macro2::{Ident, TokenStream, TokenTree};
use proc_macro_error::proc_macro_error;
use quote::quote;

#[proc_macro]
#[proc_macro_error]
pub fn html(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    expand(input.into()).into()
}

#[proc_macro]
#[proc_macro_error]
pub fn html_debug(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let expr = expand(input.into());
    println!("expansion:\n{}", expr);
    expr.into()
}

fn expand(input: TokenStream) -> TokenStream {
    // TODO: call `proc_macro2::Span::mixed_site()` directly when Rust 1.45 is stable
    let output_ident = TokenTree::Ident(Ident::new(
        "__maud_output",
        proc_macro::Span::mixed_site().into(),
    ));
    // Heuristic: the size of the resulting markup tends to correlate with the
    // code size of the template itself
    let size_hint = input.to_string().len();
    let markups = parse::parse(input);
    let stmts = generate::generate(markups, output_ident.clone());
    quote!({
        extern crate maud;
        let mut #output_ident = ::std::string::String::with_capacity(#size_hint);
        #stmts
        maud::PreEscaped(#output_ident)
    })
}
