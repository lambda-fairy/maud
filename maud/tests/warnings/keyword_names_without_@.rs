extern crate maud_macros;

use maud_macros::html;

fn main() {
    html! {
        if {}
    };
}
