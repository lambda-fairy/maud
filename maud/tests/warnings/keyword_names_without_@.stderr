error: found keyword `if`
 --> $DIR/keyword_names_without_@.rs:7:9
  |
7 |         if {}
  |         ^^
  |
  = help: should this be a `@if`?
